### PR TITLE
Reshape support tensor attribute

### PR DESCRIPTION
**Pull request by phlrain**
旧的方式
layers.reshape(a, [10, 1])
新的pr 支持
layers.reshape( a, [dim1, dim2])
dim1 dim2 可以是variable或者python的值

### CODE REVIEW

**Comment by luotao1** [comment]
```
[14:30:02]	140/155 Test #603: test_reshape_ngraph_op ...................***Failed    2.83 sec
[14:30:02]	test_reshape_ngraph_op failed
[14:30:02]	 ..EEEE..
[14:30:02]	======================================================================
[14:30:02]	ERROR: test_check_grad (test_reshape_op.TestReshapeOpDimInfer1)
[14:30:02]	----------------------------------------------------------------------
[14:30:02]	Traceback (most recent call last):
[14:30:02]	  File "../test_reshape_op.py", line 116, in test_check_grad
[14:30:02]	    self.check_grad(["X"], "Out")
[14:30:02]	  File "../op_test.py", line 545, in check_grad
[14:30:02]	    user_defined_grads)
[14:30:02]	  File "../op_test.py", line 590, in check_grad_with_place
[14:30:02]	    output_names, no_grad_set)
[14:30:02]	  File "../op_test.py", line 668, in _get_gradient
[14:30:02]	    executor.run(prog, feed_dict, fetch_list, return_numpy=False)))
[14:30:02]	  File "/paddle/build/python/paddle/fluid/executor.py", line 650, in run
[14:30:02]	    use_program_cache=use_program_cache)
[14:30:02]	  File "/paddle/build/python/paddle/fluid/executor.py", line 748, in _run
[14:30:02]	    exe.run(program.desc, scope, 0, True, True, fetch_var_name)
[14:30:02]	ValueError: Cannot convert the value 0xFFFFFFFFFFFFFFFF to Dimension: this value is used internally to represent a dynamic dimension.
[14:30:02]	
[14:30:02]	
```
http://ci.paddlepaddle.org/viewLog.html?buildId=107928&tab=buildLog&buildTypeId=Paddle_PrCi&logTab=tree&filter=all&_focus=13043 
@LeoZhao-Intel @baojun-nervana Could you help see how to fix this failed ut? Or enhance the ngraph  reshape op since https://github.com/PaddlePaddle/Paddle/pull/17781#issue-284277971?

File: python/paddle/fluid/tests/unittests/test_reshape_op.py [hunk 1]
```diff
@@ -82,5 +84,76 @@ def test_check_grad(self):
...
+        self.infered_shape = (5, 10)
+
+    def test_check_output(self):
+        self.check_output(no_check_set=['XShape'])
+
+    def test_check_grad(self):
+        self.check_grad(["X"], "Out")
+
+
+class TestReshapeOpDimInfer1(TestReshapeOp_attr_tensor):
```
Comment (baojun-nervana) [inline python/paddle/fluid/tests/unittests/test_reshape_op.py]: Do you have to reuse the same name TestReshapeOpDimInfer1? Can you use a different one?

File: python/paddle/fluid/tests/unittests/test_reshape_op.py [hunk 2]
```diff
@@ -82,5 +84,76 @@ def test_check_grad(self):
...
+    def init_data(self):
+        self.ori_shape = (5, 10)
+        self.new_shape = (5, -1, 5)
+        self.infered_shape = (5, -1, 5)
+
+
+class TestReshapeOpDimInfer2(TestReshapeOp_attr_tensor):
```
Comment (baojun-nervana) [inline python/paddle/fluid/tests/unittests/test_reshape_op.py]: Can you use a different test name?

File: python/paddle/fluid/tests/unittests/test_reshape_op.py [hunk 3]
```diff
@@ -82,5 +84,76 @@ def test_check_grad(self):
...
+    def init_data(self):
+        self.ori_shape = (2, 2, 6)
+        self.new_shape = (2, 0, 3, -1)
+        self.infered_shape = (2, 2, 3, -1)
+
+
+class TestReshapeOpWithInputShape(OpTest):
```
Comment (baojun-nervana) [inline python/paddle/fluid/tests/unittests/test_reshape_op.py]: Can you use a different name?

File: paddle/fluid/operators/reshape_op.cc [hunk 4]
```diff
@@ -126,9 +164,15 @@ class ReshapeOpMaker : public framework::OpProtoAndCheckerMaker {
              "the shape attribute, while the shape attribute still should be "
              "set correctly to gurantee shape inference in compile time.")
         .AsDispensable();
+    AddInput("shape_tensor",
+             "(Tensor<int32>, optional). If provided, reshape will use this"
+             "it have top prioroty than Input(Shape) and attr(shape)")
```
Comment (chengduoZH) [inline paddle/fluid/operators/reshape_op.cc]: - >  it has the highest priority compare with Input(Shape) and attr(shape).

File: paddle/fluid/operators/reshape_op.cc [hunk 5]
```diff
@@ -126,9 +164,15 @@ class ReshapeOpMaker : public framework::OpProtoAndCheckerMaker {
              "the shape attribute, while the shape attribute still should be "
              "set correctly to gurantee shape inference in compile time.")
         .AsDispensable();
+    AddInput("shape_tensor",
```
Comment (chengduoZH) [inline paddle/fluid/operators/reshape_op.cc]: `shape_tensor ` - > `ShapeTensor `

File: paddle/fluid/operators/reshape_op.cc [hunk 6]
```diff
@@ -126,9 +164,15 @@ class ReshapeOpMaker : public framework::OpProtoAndCheckerMaker {
...
+             "(Tensor<int32>, optional). If provided, reshape will use this"
+             "it have top prioroty than Input(Shape) and attr(shape)")
+        .AsDuplicable()
```
Comment (chengduoZH) [inline paddle/fluid/operators/reshape_op.cc:175]: `AsDuplicable()` ?

Comment (phlrain) [reply paddle/fluid/operators/reshape_op.cc:175]: shape 参数可以是一个列表，所以需要duplicate

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -6711,7 +6712,12 @@ def reshape(x, shape, actual_shape=None, act=None, inplace=False, name=None):
 
     # Validate the shape
     unk_dim_idx = -1
+    contain_var = False
     for dim_idx, dim_size in enumerate(shape):
+        if isinstance(dim_size, Variable):
+            contain_var = True
+            continue
```
Comment (chengduoZH) [inline python/paddle/fluid/layers/nn.py:6719]: Add unit test for this case?

File: paddle/fluid/operators/reshape_op.cc (hunk 4 above)
Comment (phlrain) [reply paddle/fluid/operators/reshape_op.cc]: shape tensor只有在layers.reshape( tensor， shape_dim), 的 shape_dim 中有var的时候才会出现，attr(shape)就不会设置值， Input(Shape) 是一种不规范的用法，新的接口能够完全支持，这个接口会逐渐放弃

**Review by chengduoZH — APPROVED** [review]

**Pull request merged**